### DISCUSSION TITLE
Uplot Integration

### DISCUSSION THREAD

**Discussion by CSharpDummy** [zauberzeug/nicegui#2669] [Ideas / Feature Requests]
Uplot is a small and VERY fast plotting library. They have a JavaScript and vue version. How difficult is it to wrap it in a class like plotly or Echart?

**Comment by falkoschindler** [comment]
Hi @CSharpDummy, thanks for the suggestion!
The API looks similar to ECharts, so I guess an it can be integrated rather smoothly.

**Comment by NvNielen** [comment]
Hello, I came across this post some time ago and gave it a go. Loving how easy it is to implement these sort of things in NiceGUI :)
I needed something that was able to handle large streams of data. In my case, I was working with software for scientific instruments, such as Zurich Instrument's MFLI, which can sample at ~100 kSa/s. 

First, I tried plotly + tsdownsampler, but I noticed that it couldn't quite keep up at higher refresh rates, even at a smaller downsample size (~1K points). Additionally, I was running it in a docker container and plotly seemed to chug a lot of RAM. So, I ended up making a uplot element with a couple of plugins (Zoom scroll + Legend tooltip) and the result was quite satisfactory:

https://github.com/user-attachments/assets/f49724f0-dbf8-4842-84b2-54d9906adff2

Streaming at around 100 kSa/s ( not a very steady stream ;-) ) and a 1M point buffer downsampled to 1K.

Pros:
- Very fast, It can handle higher refresh rates (in this demo 10Hz)
- It obviously leaves a much smaller memory footprint compared to plotly
- Quite feature rich, with many plugins available
- Can be integrated with tsdownsample as in #5704

Cons:
- Not as well maintained, therefore contains some bugs
- It does not sanitize input for performance, so for example log/log plotting with values <= 0 leads to crashes
- Works best in Chromium-based browsers, and requires some [rendering pipeline unclogging](https://github.com/leeoniya/uPlot#unclog-your-rendering-pipeline) for best performance
- Documentation is not ideal, and many plugins require some additional tinkering

Of course for now this is very much glued-together, but I think it has potential and would be a nice addition to NiceGUI

I would be interested to know what you think about this @falkoschindler

**Reply by falkoschindler** [reply]
Hi @NvNielen, thanks for sharing this — the demo looks really impressive! 100 kSa/s with a 1M point buffer downsampled to 1K at 10 Hz refresh is exactly the kind of scenario where Plotly and ECharts start to struggle, so it's great to see uPlot filling that gap.

The wrapper for something like this would be fairly thin, and the maintenance burden should be manageable — similar to how we handle other charting libraries. The uPlot quirks you mentioned (log/log crashes, Chromium-specific optimizations) are upstream issues rather than something we'd own.

Would you be interested in opening a PR? That would give us a concrete basis to evaluate the integration and discuss details like API surface, plugin handling, and documentation. Happy to help with guidance on the element wrapping patterns — have a look at how existing elements in `nicegui/elements/` are structured for reference.

**Reply by NvNielen** [reply]
Hello, thank you for sharing your thoughts on this! I definitely agree that some of the quirks would be upstream issues, though some will inevitably be there depending on the features and plugins present.

I would certainly be interested in giving it a go. I'll try to set up a base element and then open a PR, then we can indeed evaluate from there. Guidance is very much appreciated!

**Reply by evnchn** [reply]
There is our contribution guide https://github.com/zauberzeug/nicegui/blob/main/CONTRIBUTING.md, and I am happy to answer any questions you may have on top. 

The guide is a tad long but maybe you can AI skim through it.

**Comment by evnchn** [comment]
@NvNielen how does https://github.com/zauberzeug/nicegui/pull/5704 sound, though?
